consult-org.el

```diff
@@ -0,0 +1,167 @@
+;;; consult-org.el --- Consult commands for org-mode -*- lexical-binding: t -*-
+
+;; Author: Daniel Mendler and Consult contributors
+;; Maintainer: Daniel Mendler <mail@daniel-mendler.de>
+;; Created: 2020
+;; License: GPL-3.0-or-later
+;; Version: 0.6
+;; Package-Requires: ((consult "0.6") (org "9.1") (emacs "26.1"))
+;; Homepage: https://github.com/minad/consult
+
+;; This file is not part of GNU Emacs.
+
+;; This program is free software: you can redistribute it and/or modify
+;; it under the terms of the GNU General Public License as published by
+;; the Free Software Foundation, either version 3 of the License, or
+;; (at your option) any later version.
+
+;; This program is distributed in the hope that it will be useful,
+;; but WITHOUT ANY WARRANTY; without even the implied warranty of
+;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+;; GNU General Public License for more details.
+
+;; You should have received a copy of the GNU General Public License
+;; along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+;;; Commentary:
+
+;; Provides a completing-read interface for some Org mode
+;; actions. This is an extra package, to allow lazy loading of Org.
+
+;;; Code:
+
+(require 'consult)
+(require 'org-agenda)
+(require 'org-clock)
+
+(defvar consult-org--history nil)
+
+(defun consult-org--narrow ()
+  "Narrowing configuration for `consult-org' commands."
+  (let ((todo-keywords (seq-filter
+                        (lambda (it) (<= ?a (car it) ?z))
+                        (mapcar (lambda (s)
+                                  (pcase-let ((`(,a ,b) (split-string s "(")))
+                                    (cons (downcase (string-to-char (or b a))) a)))
+                                (mapcan 'cdr org-todo-keywords)))))
+    (cons (lambda (cand)
+            (cond ((<= ?1 consult--narrow ?9)
+                   (<= (get-text-property 0 'consult-org--level cand)
+                       (- consult--narrow ?0)))
+                  ((<= ?A consult--narrow ?Z)
+                   (eq (get-text-property 0 'consult-org--priority cand)
+                       consult--narrow))
+                  ((when-let ((todo (alist-get consult--narrow todo-keywords)))
+                     (string-equal (get-text-property 0 'consult-org--todo cand)
+                                   todo)))))
```

Comment:
I would like to see the narrowing function to be simplified. You can either encode the possible keys into an integer as in `consult-minor-mode-menu`: 

https://github.com/minad/consult/blob/f22309220abbc77074eddd16530d98e268695105/consult.el#L3299-L3302

Alternatively store the possible narrow keys as a string in each candidate and then check if `consult--narrow` is an element of this string.

Reply:
Is the problem that this lambda isn't sufficiently elegant? I agree it's ugly, but I think your suggestion would spread the inelegance to other places (e.g., `todo-keywords` would have to be available in `consult-org--entries` as well).

Reply:
Yes, my problem here is that the narrow function is overly complicated. I prefer to push the complexity into the candidate generation. I think overall it will be a bit simpler.